duckduckgo/macos-browser
# Support for reordering Favorites

Task/Issue URL: https://app.asana.com/0/1177771139624306/1202117615001669/f
Tech Design URL: https://app.asana.com/0/481882893211075/1203352064964054/f

**Description**:
This change updates how Favorites are stored in Core Data. A dedicated Favorites Folder is now created.
This folder is outside of the Bookmarks tree. Individual bookmarks are marked as favorite by creating
a relationship to this folder (the relationship is called `favoritesFolder` and is separate from `parentFolder`).
The relationship on the Favorites Folder side is ordered, which is how favorites order is persisted.

BookmarkStore and BookmarkManager get `moveFavorites` API which is sibling to `move(objectUUIDs:)` but
works on favorites relationship (without affecting Bookmarks folder tree).

Reordering is available:
* from Bookmarks Management on all operating systems
* from the Home Page favorites view on macOS Big Sur and newer (the view uses LazyVGrid introduced in BigSur).

Heavyweight database migration was added to update `isFavorite` Core Data attribute to the relationship.



**Steps to test this PR**:
1. **Run the app from develop branch** (or from any branch other than the one for this PR).
1. Make sure you've added some bookmarks and favorites.
1. Run the app from this PR's branch and verify that your bookmarks and favorites are intact.
1. Verify that reordering favorites on Home Page works.
    1. Check with:
        1.  fewer than 5 favorites
        1. 5 (or multiple of 6 minus 1) favorites
        1. multiple of 6 favorites
    1. After you rearrange favorites, open another home tab and verify that the order of favorites on both tabs is identical
    1. Open Bookmarks Management view and verify that as you reorder favorites on the home page, the new order is applied in Bookmarks Management favorites list (note that it the database is saved only on drag ended event, not during dragging).
1. Verify that reordering favorites in the Bookmarks Management view works:
    1. Check by reordering single favorite and by selecting multiple and dragging them down or up

**Testing checklist**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**
